### PR TITLE
branchprotector checks if updates are needed

### LINKED ISSUE #11989: Make branchprotector check if updates are needed applying branch protection policies.
**What would you like to be added**:
`branchprotector` currently calls `UpdateBranchProtection` for every branch even if the branches are already up to date. It should instead fetch the current branch protection policy and see if updates are needed before calling `UpdateBranchProtection`.

**Why is this needed**:
This should dramatically improve the performance of `branchprotector` and ultimately save API tokens because the `ghproxy` cache will cache the branch policy lookups, fulfilling subsequent requests for free and without client throttling. The number of `UpdateBranchProtection` calls should also be greatly reduced because very little (if anything) changes between most `branchprotector` runs.

/assign @fejta 
/area prow/branchprotector

### PR DESCRIPTION
**Pull request by adshmh**
branchprotector now fetches the current protection object of a branch from the github API, and matches it against the configured policy. An update is requested only if the two do not match. Closes #11989 .

### CODE REVIEW

File: prow/github/types.go [hunk 1]
```diff
@@ -357,6 +357,16 @@ type Branch struct {
...
 
+// BranchProtection represents protections
+// in place for a branch
+// See also: https://developer.github.com/v3/repos/branches/#get-branch-protection
+type BranchProtection struct {
+	RequiredStatusChecks       *RequiredStatusChecks       `json:"required_status_checks"`
+	EnforceAdmins              *bool                       `json:"enforce_admins"`
+	RequiredPullRequestReviews *RequiredPullRequestReviews `json:"required_pull_request_reviews"`
+	Restrictions               *Restrictions               `json:"restrictions"`
+}
```
Comment (cjwagner) [inline prow/github/types.go]: This struct doesn't appear to match the [payload example](https://developer.github.com/v3/repos/branches/#response-2). In particular the `EnforceAdmins` field.

Are these fields always present in the payload? If so we should use struct types instead of pointers to structs.

File: prow/github/client.go [hunk 2]
```diff
@@ -1568,6 +1569,23 @@ func (c *client) GetBranches(org, repo string, onlyProtected bool) ([]Branch, er
...
 
+// GetBranchProtection returns current protection object for the branch
+//
+// See https://developer.github.com/v3/repos/branches/#get-branch-protection
+func (c *client) GetBranchProtection(org, repo, branch string) (*BranchProtection, error) {
+	c.log("GetBranchProtection", org, repo, branch)
+	var retProtection BranchProtection
+	_, err := c.request(&request{
+		method:    http.MethodGet,
+		path:      fmt.Sprintf("/repos/%s/%s/branches/%s/protection", org, repo, branch),
```
Comment (cjwagner) [inline prow/github/client.go]: This endpoint appears to return 404s if the branch has no protection configured which will cause the function to incorrectly return an error. Unfortunately the endpoint also appears to return a 404 if the access token used by the request does not have sufficient privileges so this function will also need to check if the message contains `Branch not protected` to distinguish between legitimate errors and no protection configured.

File: prow/cmd/branchprotector/protect.go [hunk 3]
```diff
@@ -289,3 +310,34 @@ func (p *protector) UpdateBranch(orgName, repo string, branchName string, branch
...
+	}
+
+	if bp1.Restrictions != nil {
+		sort.Strings(*bp1.Restrictions.Teams)
+		sort.Strings(*bp1.Restrictions.Users)
+	}
+	if bp2.Restrictions != nil {
+		sort.Strings(*bp2.Restrictions.Teams)
+		sort.Strings(*bp2.Restrictions.Users)
+	}
```
Comment (cjwagner) [inline prow/cmd/branchprotector/protect.go]: nit: Instead of duplicating all the sorting logic, you could move it to an anonymous function. e.g.
```go
sortBP := func(bp *github.BranchProtection) {
  // sort bp
}
sortBP(&bp1)
sortBP(&bp2)
```

File: prow/cmd/branchprotector/protect_test.go [hunk 4]
```diff
@@ -126,6 +127,14 @@ func (c fakeClient) GetBranches(org, repo string, onlyProtected bool) ([]github.
 	return b, nil
 }
 
+func (c *fakeClient) GetBranchProtection(org, repo, branch string) (*github.BranchProtection, error) {
+	ctx := org + "/" + repo + "=" + branch
+	if bp, ok := c.branchProtections[ctx]; ok {
+		return &bp, nil
+	}
+	return nil, nil
```
Comment (cjwagner) [inline prow/cmd/branchprotector/protect_test.go]: The behavior for this case (no branch protection configured) doesn't match the behavior of the real client against the GitHub API since the API returns a 404 which the client considers an error.

File: prow/cmd/branchprotector/protect.go [hunk 5]
```diff
@@ -276,11 +279,29 @@ func (p *protector) UpdateBranch(orgName, repo string, branchName string, branch
 		logrus.Infof("%s/%s=%s: already unprotected", orgName, repo, branchName)
 		return nil
 	}
+
+	currentBP, err := p.client.GetBranchProtection(orgName, repo, branchName)
```
Comment (cjwagner) [inline prow/cmd/branchprotector/protect.go]: It would be preferable if we could list all the branch protections settings for a repo instead of requesting them individually based on Prow's branch protector config. Listing would save API tokens and more importantly, allow us to identify protection settings not explicitly mentioned in Prow's config that should be pruned.

Could you please evaluate if [this API endpoint](https://developer.github.com/v3/repos/branches/#list-branches) (which is already called [here](https://github.com/kubernetes/test-infra/blob/87aa9f5b3a9f96526e460810f360a5fd5afa6e7d/prow/cmd/branchprotector/protect.go#L250)) or any other provides enough fields to work as a cheaper alternative?

Comment (adshmh) [reply prow/cmd/branchprotector/protect.go]: Thanks for pointing this out. From my tests, it seems the [Get branch list endpoint](https://developer.github.com/v3/repos/branches/#list-branches) and [Get a single branch endpoint](https://developer.github.com/v3/repos/branches/#get-branch) both return a partially filled protection object. 
I looked at all available API endpoints, and did not find any other options for listing all branch protections either (perhaps I can make an enhancement request to github support?)

File: prow/github/types.go (hunk 1 above)
Comment (adshmh) [reply prow/github/types.go]: Thank you for the review. From my tests it seems all fields, except `EnforceAdmins`, can be absent. I will update the PR to address this.

File: prow/cmd/branchprotector/protect.go (hunk 5 above)
Comment (cjwagner) [reply prow/cmd/branchprotector/protect.go]: Yeah opening an enhancement request ticket with GitHub support sounds like a good idea.
In the meantime if we are going to use the per branch `GetBranchProtection` calls please add a comment explaining why we have to do that instead of listing the protection for multiple branches at once.

File: prow/cmd/branchprotector/protect.go (hunk 3 above)
Comment (adshmh) [reply prow/cmd/branchprotector/protect.go]: Thank you for the code review. This reads much better. I will update the PR.

File: prow/github/client.go (hunk 2 above)
Comment (adshmh) [reply prow/github/client.go]: Thanks, I will update the function to differentiate 404 due to `Branch not protected` and 404 due to any other reason.

File: prow/cmd/branchprotector/protect_test.go (hunk 4 above)
Comment (adshmh) [reply prow/cmd/branchprotector/protect_test.go]: Thank you for the review. I will update this to match the new behaviour, i.e. empty branch protection object when branch protection is not configured.

File: prow/cmd/branchprotector/protect.go (hunk 5 above)
Comment (adshmh) [reply prow/cmd/branchprotector/protect.go]: I have updated the PR. I also sent a request to github support. Will let you know once I hear back from them.

File: prow/cmd/branchprotector/protect.go [hunk 6]
```diff
@@ -276,11 +279,32 @@ func (p *protector) UpdateBranch(orgName, repo string, branchName string, branch
...
+		policyBP = github.BranchProtection{
+			EnforceAdmins:              r.EnforceAdmins != nil && *r.EnforceAdmins,
+			RequiredPullRequestReviews: r.RequiredPullRequestReviews,
+			RequiredStatusChecks:       r.RequiredStatusChecks,
+			Restrictions:               r.Restrictions,
+		}
 	}
+
+	if req != nil && equalBranchProtections(*currentBP, policyBP) {
+		logrus.Infof("%s/%s=%s: current branch protection matches policy, skipping", orgName, repo, branchName)
```
Comment (cjwagner) [inline prow/cmd/branchprotector/protect.go]: Please change this to the debug level.
```suggestion
		logrus.Debugf("%s/%s=%s: current branch protection matches policy, skipping", orgName, repo, branchName)
```

File: prow/github/client.go [hunk 7]
```diff
@@ -1568,6 +1569,51 @@ func (c *client) GetBranches(org, repo string, onlyProtected bool) ([]Branch, er
...
+
+	ge := &githubError{}
+	if err := json.Unmarshal(body, ge); err != nil {
+		return nil, err
+	}
+
+	// If the error was because the branch is not protected, we return an
+	// empty BranchProtection object
+	if ge.Message == "Branch not protected" {
+		return &BranchProtection{}, nil
```
[the remark this answers is not in the document]
Comment (adshmh) [reply prow/github/client.go]: Yes, I used an empty object first, but after a review using nil seems better as there won't be a need to update the unit tests and  an `UpdateBranchProtection`  API call with all parameters set to nil sets the `protected` flag on the branch.

File: prow/cmd/branchprotector/protect.go (hunk 6 above)
Comment (adshmh) [reply prow/cmd/branchprotector/protect.go]: Thanks, will update the PR.

File: prow/github/types.go (hunk 1 above)
Comment (adshmh) [reply prow/github/types.go]: I see, fixed it. `enforce_admins` is always present in the payload in my tests, so will use a struct. All other fields can be missing from the payload.

File: prow/cmd/branchprotector/protect.go (hunk 5 above)
Comment (adshmh) [reply prow/cmd/branchprotector/protect.go]: Received a response from github support, saying they have added my +1 for this to their internal feature list (perhaps the `test-infra` team may consider requesting the same feature assuming  enhancement requests carry different weights depending on support level to which the repository/team has access).

File: prow/cmd/branchprotector/protect_test.go (hunk 4 above)
Comment (adshmh) [reply prow/cmd/branchprotector/protect_test.go]: Fixing the client to return `nil` if the branch is not protected should fix this as well.

**Comment by stevekuznetsov** [comment]
Super excited for this one :)

File: prow/github/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (cjwagner) [reply prow/github/client.go]: > an `UpdateBranchProtection` API call with all parameters set to nil sets the `protected` flag on the branch.

Oh, that definitely makes differentiating empty from nil a requirement. 👍

File: prow/cmd/branchprotector/protect.go [hunk 8]
```diff
@@ -276,11 +279,34 @@ func (p *protector) UpdateBranch(orgName, repo string, branchName string, branch
...
+			RequiredPullRequestReviews: r.RequiredPullRequestReviews,
+			RequiredStatusChecks:       r.RequiredStatusChecks,
+			Restrictions:               r.Restrictions,
+		}
 	}
+
+	if req != nil && currentBP != nil && equalBranchProtections(*currentBP, policyBP) {
+		logrus.Debugf("%s/%s=%s: current branch protection matches policy, skipping", orgName, repo, branchName)
+		return nil
+	}
```
Comment (cjwagner) [inline prow/cmd/branchprotector/protect.go]: I'm pretty sure we can also skip if `req == nil && currentBP == nil`. That will prevent unnecessary calls to `RemoveBranchProtection` here: https://github.com/kubernetes/test-infra/blob/021b9b2aea9b9254d38d1a5481bf7ff4108d8cad/prow/cmd/branchprotector/protect.go#L156

File: prow/cmd/branchprotector/protect_test.go [hunk 9]
```diff
@@ -697,6 +707,61 @@ branch-protection:
...
+							Teams: &[]string{"oncall", "sres"},
+						},
+					},
+					Restrictions: &github.Restrictions{
+						Users: &[]string{"cindy"},
+						Teams: &[]string{"config-team"},
+					},
+				},
+			},
+		},
```
Comment (cjwagner) [inline prow/cmd/branchprotector/protect_test.go]: It would be nice to have a test case for "make request if branch protection is present, but out of date" and a test case for "don't remove branch protection if protection is not present and config sets `protect: false`" (which wont work until the `req == nil && currentBP == nil` case is skipped).

File: prow/cmd/branchprotector/protect.go (hunk 8 above)
Comment (adshmh) [reply prow/cmd/branchprotector/protect.go]: Thanks for pointing that out. While reviewing the existing unit tests, seems like that particular case, i.e. currently unprotected branch and policy sets protected to false, is already covered:

https://github.com/kubernetes/test-infra/blob/ba215daf8be54c80be6be4c66503b1a5e60d09c9/prow/cmd/branchprotector/protect_test.go#L678-L688

and 

https://github.com/kubernetes/test-infra/blob/ba215daf8be54c80be6be4c66503b1a5e60d09c9/prow/cmd/branchprotector/protect.go#L275-L278

I think that's because the github API's reponse to `GetBranches` has a `protected` field in the payload, so the existing code uses it for that particular optimization.

Please let me know if this is not correct and the `req == nil && currentBP == nil` optimisation is needed .

File: prow/cmd/branchprotector/protect_test.go (hunk 9 above)
Comment (adshmh) [reply prow/cmd/branchprotector/protect_test.go]: Thank you for the review. Added the first test case. Will hold off on the second test case to see if the existing optimisation and unit tests already cover it.

File: prow/github/client.go [hunk 10]
```diff
@@ -1568,6 +1569,51 @@ func (c *client) GetBranches(org, repo string, onlyProtected bool) ([]Branch, er
...
+	}
+
+	ge := &githubError{}
+	if err := json.Unmarshal(body, ge); err != nil {
+		return nil, err
+	}
+
+	// If the error was because the branch is not protected, we return a
+	// nil pointer to indicate this.
+	if ge.Message == "Branch not protected" {
```
Comment (fejta) [inline prow/github/client.go]: Relying on error strings is generally ill advised as it is brittle. Are you sure this is worth it?

Comment (adshmh) [reply prow/github/client.go]: Thank you for the review. The github API for getting branch protection returns 404 for both a) unprotected branch, and b) authorisation error (perhaps for other `Not Found` errors as well). The only way to distinguish these seems to be looking at the error message body, which we need since `branchprotector` does not consider the former, i.e. unprotected branch, an error.

File: prow/github/client.go [hunk 11]
```diff
@@ -1568,6 +1569,51 @@ func (c *client) GetBranches(org, repo string, onlyProtected bool) ([]Branch, er
...
+		// - The access token used does not have sufficient privileges
+		// We therefore need to introspect the response body.
+		exitCodes: []int{200, 404},
+	})
+
+	switch {
+	case err != nil:
+		return nil, err
+	case code == 200:
+		bp := &BranchProtection{}
```
[the remark this answers is not in the document]
Comment (adshmh) [reply prow/github/client.go]: Thanks, I will update the PR.

File: prow/cmd/branchprotector/protect.go [hunk 12]
```diff
@@ -21,6 +21,8 @@ import (
 	"flag"
 	"fmt"
 	"os"
+	"reflect"
```
[the remark this answers is not in the document]
Comment (adshmh) [reply prow/cmd/branchprotector/protect.go]: Thank you for the review. Do you mean to use a type-specific equality method instead?

[the remark this answers is not in the document]
Comment (fejta) [reply prow/cmd/branchprotector/protect.go]: Yes, you can either write it manually or else use something like https://godoc.org/k8s.io/gengo/examples/deepcopy-gen

File: prow/cmd/branchprotector/protect.go (hunk 3 above)
Comment (adshmh) [reply prow/cmd/branchprotector/protect.go]: Since this was asked as a question, I have left as-is pending a decision form maintainers. Please let me know if the casing of strings needs to be considered as well.

File: prow/cmd/branchprotector/protect.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (adshmh) [reply prow/cmd/branchprotector/protect.go]: Thanks, implemented a type-specific equality function for this one.

File: prow/github/types.go (hunk 1 above)
Comment (stevekuznetsov) [reply prow/github/types.go]: nit: would want `omitempty` on optional fields for serialization
